storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,22 +3580,413 @@ dberr_t fsp_sys_tablespace_validate()
+    err= DB_CORRUPTION;
+
+  block->page.unfix();
+
+  if (err)
+    return err;
+
+  block= buf_pool.page_fix(page_id_t{0, DICT_HDR_PAGE_NO}, &err,
+                           buf_pool_t::FIX_WAIT_READ);
+  if (!block)
+    return err;
+
+  if (!inodes->insert_seg(block->page.frame + DICT_HDR + DICT_HDR_FSEG_HEADER))
+    err= DB_CORRUPTION;
+
+  block->page.unfix();
+
+  if (err)
+    return err;
+
+  block= buf_pool.page_fix(page_id_t{0, FSP_IBUF_HEADER_PAGE_NO},
+                           &err, buf_pool_t::FIX_WAIT_READ);
+  if (!block)
+    return err;
+  if (!inodes->insert_seg(block->page.frame + PAGE_DATA))
+    err= DB_CORRUPTION;
+
+  block->page.unfix();
+
+  /* Get rollback segment header page */
+  for (ulint rseg_id= 0; rseg_id < TRX_SYS_N_RSEGS && err == DB_SUCCESS;
+       rseg_id++)
+  {
+    trx_rseg_t *rseg= &trx_sys.rseg_array[rseg_id];
+    if (rseg->space->id == 0)
+    {
+      block= buf_pool.page_fix(rseg->page_id(), &err,
+                               buf_pool_t::FIX_WAIT_READ);
+      if (!block)
+        break;
+      if (!inodes->insert_seg(block->page.frame + TRX_RSEG +
+                              TRX_RSEG_FSEG_HEADER))
+        err= DB_CORRUPTION;
+      block->page.unfix();
+
+      /* Even after slow shutdown, there is a possiblity that
+      cached undo log can exist. So store the segment as used one */
+      for (trx_undo_t *undo= UT_LIST_GET_FIRST(rseg->undo_cached);
+           undo && err == DB_SUCCESS;
+           undo= UT_LIST_GET_NEXT(undo_list, undo))
+      {
+        block= buf_pool.page_fix(page_id_t{0, undo->hdr_page_no}, &err,
+                                 buf_pool_t::FIX_WAIT_READ);
+        if (!block)
+	  return err;
+	if (!inodes->insert_seg(block->page.frame + TRX_UNDO_SEG_HDR +
+                                TRX_UNDO_FSEG_HEADER))
+          err= DB_CORRUPTION;
+        block->page.unfix();
+      }
+    }
+  }
+
+  if (err == DB_SUCCESS)
+    err= fsp_table_inodes(inodes, mtr);
+  return err;
+}
+
+/** Free the extents, fragment page from the given inode
+@param page_no index node page number
+@param offset  index node offset within page
+@return error code */
+static dberr_t fseg_inode_free(uint32_t page_no, uint16_t offset)
+{
+  fil_space_t *space= fil_system.sys_space;
+  dberr_t err= DB_SUCCESS;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  buf_block_t *iblock= buf_page_get_gen(page_id_t{0, page_no}, 0,
+                                        RW_X_LATCH, nullptr, BUF_GET,
+                                        &mtr, &err);
+
+  fseg_inode_t *inode= nullptr;
+  DBUG_EXECUTE_IF("unused_undo_free_fail_4", iblock= nullptr;);
+  if (!iblock)
+    goto func_exit;
+
+  sql_print_information("InnoDB: Freeing the unused segment in page "
+                        UINT32PF, page_no);
+  inode= iblock->page.frame + offset;
+  err= fseg_free_step_low(iblock, inode, space, &mtr, nullptr, true);
+  while (err == DB_SUCCESS_LOCKED_REC)
```

Comment:
It looks like this call could be made inside the `while` block, to avoid duplicating the call at the end of the `while` body:
```c++
  while ((err= fseg_free_step_low(…) == DB_SUCCESS_LOCKED_REC)
```
I would like to see a source code comment after the `while` loop body, before `func_exit:` that would explain why there is no call to `fseg_free_step()`, like we have in `trx_purge_free_segment()`. Is the header page being freed as well? If not, what will remain in the header page?